packages/playwright-core/src/server/page.ts

```diff
@@ -903,7 +903,21 @@ function addPageBinding(playwrightBinding: string, bindingName: string, needsHan
       }
       payload = { name: bindingName, seq, serializedArgs };
     }
+
+    // if any of these types defines toJSON, it can mess up JSON.stringify
+    const toJSONs: [Object | Array<any> | Number | Boolean | String, any][] = [];
+    for (const ctor of [Object, Array, Number, Boolean, String]) {
+      if ('toJSON' in ctor.prototype) {
+        toJSONs.push([ctor, (ctor.prototype as any).toJSON]);
+        delete ctor.prototype.toJSON;
```

Comment:
I don't think we want to mess up with userland overrides. Can we have any other workaround instead?

Reply:
I restore them after `JSON.stringify` (and my test already checks it) but I suppose there are some corner cases that I'm not thinking of. One option would be to only rely on `JSON.stringify` if none of those types defines `toJSON`, otherwise implement it ourselves.

Reply:
There's probably a simple solution: ensure all properties of `SerializableValue` override `toJSON` as `null` if it's prototype defines it. I'll test it today.

Reply:
That's an interesting idea. Did you have a chance to try it out?

Reply:
I just force pushed a new patch. I implemented a custom "JSON.stringify" instead because it adding `toJSON` property to objects could lead to side effects.

It is working fine with `Array.prototype.toJSON`.

Primitive types serialization like number or string are not impacted with a custom `toJSON` implementation so I left them out.

I also left out `Object.prototype.toJSON`. If users mess up this function, it will make playwright fail, but I don't think there's much we can do, because it's a fallback mechanism for custom object serializattion:

https://github.com/microsoft/playwright/blob/8995ace8253ad25865a04452408967e929911205/packages/playwright-core/src/server/isomorphic/utilityScriptSerializers.ts#L214-L224

Reply:
I looked at the last commit, and I don't think it's worth it. Manually rolling `JSON.stringify` is non-trivial. For example, your current implementation ignores `toJSON()` for objects and does not handle infinite recursion. I am sure there are more edge cases that built-in implementation covers.

I think it's ok if `exposeBinding` does not work on some sites that mess up with builtins. To be testable, these will have to be updated. Alternatively, one can remove custom `toJSON()` override from `addInitScript()`, as long as the site keeps working.

Reply:
Just a note: my implementation is not intended to be generic, it doesn't handle recursion simply because the structure is not recursive (it's an object with `SerializableValue` objects). But I see your point

Reply:
For the record, we had a similar attempt (in spirit) to make codegen work on sites that mess up with builtins - see #6187. However, we have decided that's too much complexity for too little benefit, and it worked so far. This case seems similar.

Reply:
I agree, sometimes it's not worth it. I would propose a change however: this problem manifests itself as a `serializableArgs.map is not a function`, just like in #22460, from 

https://github.com/microsoft/playwright/blob/8995ace8253ad25865a04452408967e929911205/packages/playwright-core/src/server/page.ts#L849

If `serializedArgs` is not an array, it should throw a better error to the end user, so that they know Array.prototype.toJSOn is probably messed up.

Reply:
That sounds good to me, looking forward to the PR.

Reply:
I just pushed a new change, with a more explicit error message.